File: assets/js/react/utils/linkify.ts
export type LinkifiedSegment =
  | { kind: "text"; value: string }
  | { kind: "link"; label: string; href: string }

// Hostname with a letter-only TLD (avoids matching version numbers like 1.2.3).
const DOMAIN_HOST =
  String.raw`[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?(?:\.[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?)*\.[a-zA-Z]{2,}`
const DOMAIN = String.raw`(?:www\.)?${DOMAIN_HOST}`
const EMAIL = String.raw`[a-zA-Z0-9._%+-]+@${DOMAIN_HOST}`

const MARKDOWN_LINK_RE = new RegExp(
  String.raw`\[([^\]]+)\]\((https?:\/\/[^\s)]+|mailto:[^\s)]+|${EMAIL}|${DOMAIN}(?:\/[^\s)]*)?)\)`,
  "gi"
)
const BARE_URL_RE = new RegExp(
  String.raw`(?:mailto:[^\s<]+|${EMAIL}|https?:\/\/[^\s<]+|${DOMAIN}(?:\/[^\s<]*)?)`,
  "gi"
)
const TRAILING_PUNCTUATION_RE = /[.,;:!?'")\]]+$/
const HOSTNAME_RE =
  /^[a-zA-Z0-9](?:[a-zA-Z0-9-]*[a-zA-Z0-9])?(?:\.[a-zA-Z0-9](?:[a-zA-Z0-9-]*[a-zA-Z0-9])?)+$/
const EMAIL_ANCHORED_RE = new RegExp(String.raw`^${EMAIL}$`)
const MAILTO_HREF_RE = new RegExp(
  String.raw`^mailto:${EMAIL}(?:\?[a-zA-Z0-9_\-.*=&%+]*)?$`,
  "i"
)

const ALLOWED_SCHEMES = new Set(["http", "https", "mailto"])
const DISALLOWED_SCHEME_BEFORE_RE =
  /(?:ftp|file|javascript|data|vbscript):[^\s]*$/i
const MAX_MAILTO_LEN = 2048
const EXPLICIT_SCHEME_RE = /^([a-zA-Z][a-zA-Z0-9+.-]*):/
const ASCII_SCHEME_RE = /^[a-zA-Z][a-zA-Z0-9+.-]*:$/

export function sanitizeLinkInput(url: string): string {
  return url.trim().normalize("NFKC")
}

export function isMailtoHref(href: string): boolean {
  return sanitizeLinkInput(href).toLowerCase().startsWith("mailto:")
}

function explicitScheme(url: string): string | null | "invalid" {
  const match = url.match(EXPLICIT_SCHEME_RE)
  if (!match) return null
  const prefix = match[0]
  if (!ASCII_SCHEME_RE.test(prefix)) return "invalid"
  return match[1].toLowerCase()
}

function isValidMailto(url: string): boolean {
  if (url.length > MAX_MAILTO_LEN) return false
  return MAILTO_HREF_RE.test(url)
}

function isValidWebUrl(url: string): boolean {
  try {
    const parsed = new URL(/^https?:\/\//i.test(url) ? url : `https://${url}`)
    if (parsed.protocol !== "http:" && parsed.protocol !== "https:") return false
    if (!parsed.hostname.includes(".")) return false
    return HOSTNAME_RE.test(parsed.hostname)
  } catch {
    return false
  }
}

export function normalizeUrl(url: string): string {
  const sanitized = sanitizeLinkInput(url)
  if (sanitized.toLowerCase().startsWith("mailto:")) return sanitized
  if (/^https?:\/\//i.test(sanitized)) return sanitized
  if (EMAIL_ANCHORED_RE.test(sanitized)) return `mailto:${sanitized}`
  return `https://${sanitized}`
}

export function isSafeUrl(url: string): boolean {
  const sanitized = sanitizeLinkInput(url)
  if (sanitized === "") return false

  const scheme = explicitScheme(sanitized)
  if (scheme === "invalid") return false
  if (scheme && !ALLOWED_SCHEMES.has(scheme)) return false

  if (sanitized.toLowerCase().startsWith("mailto:")) {
    return isValidMailto(sanitized)
  }

  if (scheme === "http" || scheme === "https") {
    return isValidWebUrl(sanitized)
  }

  if (EMAIL_ANCHORED_RE.test(sanitized)) return true

  return isValidWebUrl(sanitized)
}

function hasDisallowedSchemePrefix(text: string, start: number): boolean {
  return DISALLOWED_SCHEME_BEFORE_RE.test(text.slice(0, start))
}

function splitBareUrls(text: string): LinkifiedSegment[] {
  const segments: LinkifiedSegment[] = []
  let lastIndex = 0

  for (const match of text.matchAll(BARE_URL_RE)) {
    const raw = match[0]
    const start = match.index ?? 0

    if (start > lastIndex) {
      segments.push({ kind: "text", value: text.slice(lastIndex, start) })
    }

    const trailing = raw.match(TRAILING_PUNCTUATION_RE)?.[0] ?? ""
    const candidate = trailing ? raw.slice(0, -trailing.length) : raw

    if (isSafeUrl(candidate) && !hasDisallowedSchemePrefix(text, start)) {
      segments.push({ kind: "link", label: candidate, href: normalizeUrl(candidate) })
      if (trailing) {
        segments.push({ kind: "text", value: trailing })
      }
    } else {
      segments.push({ kind: "text", value: raw })
    }

    lastIndex = start + raw.length
  }

  if (lastIndex < text.length) {
    segments.push({ kind: "text", value: text.slice(lastIndex) })
  }

  return segments
}

export function parseLinkifiedText(text: string): LinkifiedSegment[] {
  if (text === "") return []

  const segments: LinkifiedSegment[] = []
  let lastIndex = 0

  for (const match of text.matchAll(MARKDOWN_LINK_RE)) {
    const label = match[1]
    const url = match[2].trim()
    const start = match.index ?? 0

    if (start > lastIndex) {
      segments.push(...splitBareUrls(text.slice(lastIndex, start)))
    }

    if (isSafeUrl(url)) {
      segments.push({ kind: "link", label, href: normalizeUrl(url) })
    } else {
      segments.push({ kind: "text", value: match[0] })
    }

    lastIndex = start + match[0].length
  }

  if (lastIndex < text.length) {
    segments.push(...splitBareUrls(text.slice(lastIndex)))
  }

  return segments.length > 0 ? segments : [{ kind: "text", value: text }]
}

File: lib/storymap/linkify.ex
defmodule Storymap.Linkify do
  @moduledoc """
  Renders plain text with markdown-style `[label](url)` links and bare URLs
  (including scheme-less domains like `example.com` and email addresses) as safe HTML.
  Only emits validated link tags; all other content is HTML-escaped.
  """

  import Phoenix.HTML
  alias Storymap.URL

  @domain_host ~r/[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?(?:\.[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?)*\.[a-zA-Z]{2,}/
  @domain ~r/(?:www\.)?#{Regex.source(@domain_host)}/
  @email ~r/[a-zA-Z0-9._%+-]+@#{Regex.source(@domain_host)}/
  @markdown_link_re Regex.compile!(
                      "\\[([^\\]]+)\\]\\((https?:\\/\\/[^\\s)]+|mailto:[^\\s)]+|#{Regex.source(@email)}|#{Regex.source(@domain)}(?:\\/[^\\s)]*)?)\\)",
                      [:caseless]
                    )
  @bare_url_re Regex.compile!(
                 "(?:mailto:[^\\s<]+|#{Regex.source(@email)}|https?:\\/\\/[^\\s<]+|#{Regex.source(@domain)}(?:\\/[^\\s<]*)?)",
                 [:caseless]
               )
  @trailing_punctuation_re ~r/[.,;:!?'")\]]+$/
  @disallowed_scheme_before_re ~r/(?:ftp|file|javascript|data|vbscript):[^\s]*$/i

  @doc """
  Returns safe HTML for a description string, or an empty string for nil/blank input.
  """
  def render(nil), do: ""
  def render(""), do: ""

  def render(text) when is_binary(text) do
    parts =
      text
      |> parse_linkified_text()
      |> Enum.map(&segment_to_html/1)
      |> Enum.map(fn {:safe, inner} -> inner end)

    {:safe, parts}
  end

  defp parse_linkified_text(""), do: []

  defp parse_linkified_text(text) do
    matches = Regex.scan(@markdown_link_re, text, return: :index)

    if matches == [] do
      split_bare_urls(text)
    else
      parse_with_markdown_links(text, matches)
    end
  end

  defp parse_with_markdown_links(text, matches) do
    {segments, last_index} =
      Enum.reduce(matches, {[], 0}, fn [
                                         {match_start, match_len},
                                         {label_start, label_len},
                                         {url_start, url_len}
                                       ],
                                       {segments, last_index} ->
        label = String.slice(text, label_start, label_len)
        url = text |> String.slice(url_start, url_len) |> String.trim()

        segments =
          if match_start > last_index do
            segments ++ split_bare_urls(String.slice(text, last_index, match_start - last_index))
          else
            segments
          end

        segment =
          if URL.safe_url?(url) do
            {:link, label, URL.normalize_url(url)}
          else
            {:text, String.slice(text, match_start, match_len)}
          end

        {segments ++ [segment], match_start + match_len}
      end)

    if last_index < byte_size(text) do
      segments ++ split_bare_urls(String.slice(text, last_index..-1//1))
    else
      segments
    end
  end

  defp split_bare_urls(""), do: []

  defp split_bare_urls(text) do
    matches = Regex.scan(@bare_url_re, text, return: :index)

    if matches == [] do
      [{:text, text}]
    else
      split_bare_url_matches(text, matches)
    end
  end

  defp split_bare_url_matches(text, matches) do
    {segments, last_index} =
      Enum.reduce(matches, {[], 0}, fn [{start, len}], {segments, last_index} ->
        raw = String.slice(text, start, len)
        {candidate, trailing} = trim_trailing_punctuation(raw)

        segments =
          if start > last_index do
            segments ++ [{:text, String.slice(text, last_index, start - last_index)}]
          else
            segments
          end

        link_segments =
          if URL.safe_url?(candidate) and not disallowed_scheme_prefix?(text, start) do
            normalized = URL.normalize_url(candidate)
            base = [{:link, candidate, normalized}]
            if trailing != "", do: base ++ [{:text, trailing}], else: base
          else
            [{:text, raw}]
          end

        {segments ++ link_segments, start + len}
      end)

    if last_index < byte_size(text) do
      segments ++ [{:text, String.slice(text, last_index..-1//1)}]
    else
      segments
    end
  end

  defp trim_trailing_punctuation(raw) do
    case Regex.run(@trailing_punctuation_re, raw, return: :index) do
      [{start, len}] ->
        {String.slice(raw, 0, start), String.slice(raw, start, len)}

      _ ->
        {raw, ""}
    end
  end

  defp disallowed_scheme_prefix?(text, start) do
    prefix = String.slice(text, 0, start)
    Regex.match?(@disallowed_scheme_before_re, prefix)
  end

  defp segment_to_html({:text, value}), do: html_escape(value)

  defp segment_to_html({:link, label, href}) do
    mailto? = String.starts_with?(String.downcase(href), "mailto:")

    security_attrs =
      if mailto? do
        ""
      else
        " target=\"_blank\" rel=\"noopener noreferrer\""
      end

    {:safe,
     [
       "<a href=\"",
       Phoenix.HTML.Engine.encode_to_iodata!(href),
       "\"",
       security_attrs,
       " class=\"link link-primary\">",
       Phoenix.HTML.Engine.encode_to_iodata!(label),
       "</a>"
     ]}
  end
end
